src/EnergyPlus/UtilityRoutines.cc

```diff
@@ -503,45 +516,16 @@ namespace UtilityRoutines {
         // PURPOSE OF THIS FUNCTION:
         // Validates fuel types and sets output strings
 
-        auto const SELECT_CASE_var(UtilityRoutines::MakeUPPERCase(FuelTypeInput));
-
-        if (SELECT_CASE_var == "ELECTRICITY") {
-            FuelTypeOutput = "Electricity";
-
-        } else if (SELECT_CASE_var == "NATURALGAS") {
-            FuelTypeOutput = "NaturalGas";
-
-        } else if (SELECT_CASE_var == "DIESEL") {
-            FuelTypeOutput = "Diesel";
-
-        } else if (SELECT_CASE_var == "GASOLINE") {
-            FuelTypeOutput = "Gasoline";
-
-        } else if (SELECT_CASE_var == "COAL") {
-            FuelTypeOutput = "Coal";
-
-        } else if (SELECT_CASE_var == "FUELOILNO1") {
-            FuelTypeOutput = "FuelOilNo1";
-
-        } else if (SELECT_CASE_var == "FUELOILNO2") {
-            FuelTypeOutput = "FuelOilNo2";
-
-        } else if (SELECT_CASE_var == "PROPANE") {
-            FuelTypeOutput = "Propane";
-
-        } else if (SELECT_CASE_var == "OTHERFUEL1") {
-            FuelTypeOutput = "OtherFuel1";
-
-        } else if (SELECT_CASE_var == "OTHERFUEL2") {
-            FuelTypeOutput = "OtherFuel2";
-
+        std::string ucInputFuelType = UtilityRoutines::MakeUPPERCase(FuelTypeInput);
+        if (fuelDictionary.find(ucInputFuelType) != fuelDictionary.end()) {
```

Comment:
Just leave this code as is, and we can fix it up properly later. Don't do this.

Reply:
The right way to do this is to have `std::array<std::string_view> Names, NamesUC` translate to `enum class` once and then use that look up whichever you want.

Reply:
I'll happily revert this back, but I do wonder where you'd like to see this ultimately go.  In this particular case, there are likely more elegant solutions that involve changing the client code rather than changing the guts of these functions themselves.  But just to understand, I'd like to consider a quick hypothetical case where we want to clean this functions up in a purely isolated bubble.

Two issues I see are:
1. The two functions do slightly different things with the list of fuel types
2. The first function is only checking a subset of the full list before changing the extended values.

Would this be your preferred approach for these two functions?

```
// fuel types magically declared in one single place in the code
enum class CoreFuelType {
    Invalid = -1, Electricity, NaturalGas, Diesel, Gasoline, Coal, FuelOilNo1, FuelOilNo2, Propane, OtherFuel1, OtherFuel2, Num
};
constexpr std::array<std::string_view, static_cast<int>(CoreFuelType::Num)> coreFuelNames = {
    "Electricity", "NaturalGas", "Diesel", "Gasoline", "Coal", "FuelOilNo1", "FuelOilNo2", "Propane", "OtherFuel1", "OtherFuel2"
};
constexpr std::array<std::string_view, static_cast<int>(CoreFuelType::Num)> coreFuelNamesUC = {
    "ELECTRICITY", "NATURALGAS", "DIESEL", "GASOLINE", "COAL", "FUELOILNO1", "FUELOILNO2", "PROPANE", "OTHERFUEL1", "OTHERFUEL2"
};
enum class ExtendedFuelType {
    Invalid = -1, Steam, DistrictHeating, DistrictCooling, Num
};
constexpr std::array<std::string_view, static_cast<int>(ExtendedFuelType::Num)> extendedFuelNames = {
    "Steam", "DistrictHeating", "DistrictCooling"
};
constexpr std::array<std::string_view, static_cast<int>(ExtendedFuelType::Num)> extendedFuelNamesUC = {
    "STEAM", "DISTRICTHEATING", "DISTRICTCOOLING"
};

// then the functions
bool ValidateFuelType(
    std::string const &FuelTypeInput,
    std::string &FuelTypeOutput,
    bool const AllowSteamAndDistrict) {

    // It first searches the core list, and returns if found.
    auto fuelInputUpperCase = UtilityRoutines::MakeUPPERCase(FuelTypeInput);
    CoreFuelType fuel = static_cast<CoreFuelType>(getEnumerationValue(coreFuelNamesUC, fuelInputUpperCase));
    if (fuel != CoreFuelType::Invalid) {
        FuelTypeOutput = coreFuelNames[static_cast<int>(fuel)];
        return false;
    }

    // If it doesn't find it in the core list, it checks the steam/district flag to determine whether to check those specific fuel types
    if (AllowSteamAndDistrict) {
        ExtendedFuelType extendedFuel = static_cast<ExtendedFuelType>(getEnumerationValue(extendedFuelNamesUC, fuelInputUpperCase));
        if (extendedFuel != ExtendedFuelType::Invalid) {
            FuelTypeOutput = extendedFuelNames[static_cast<int>(extendedFuel)];
            return false;
        }
    }

    // If we make it this far, we have not found a match, so just return that it was an error
    return true;
}

bool ValidateFuelTypeWithAssignResourceTypeNum(
    std::string const &FuelTypeInput,
    std::string &FuelTypeOutput,
    DataGlobalConstants::ResourceType &FuelTypeNum) {

    // this function is expecting the fuel type in upper case already, and does not check the extended fuel list, just the core
    CoreFuelType fuel = static_cast<CoreFuelType>(getEnumerationValue(coreFuelNamesUC, FuelTypeInput));
    if (fuel != CoreFuelType::Invalid) {
        FuelTypeOutput = coreFuelNames[static_cast<int>(fuel)];
        FuelTypeNum = DataGlobalConstants::AssignResourceTypeNum(FuelTypeInput);
        return false;
    }

    // if we didn't get a match, we return an error
    return true;
}
```

An alternative would be to have two separate enum lists that each declare a full set of fuels to search, rather than a core list and an extension.  (It would be nice if we could just extend a previously declared array to build on it at compile time...).

Anyway, this is more just wondering where you'd like to see this go for the future.  I'll revert my changes unless there is something obvious I can do here without it getting wild with scope creep.

Reply:
Is this the `FuelType`/`ResourceType` issue? I would say leave it for now, and try to come up with a cleaner solution that tries to come up with a cleaner solution that maybe also involves client code. Yes, the ability to make one enumeration a subtype (i.e., extension) of another enumeration sounds like it would be helpful. I suspect it is not there because it opens up some terrible possibilities. Or maybe it's fine and they just haven't gotten around to it yet. Sans that language feature, I would suggest defining one enumeration in terms of the other, e.g., 

```cpp
enum class FuelType {
   Invalid = -1,
   Electricity,
   NaturalGas,
   ...
   Num
};

enum class ResourceType {
   Invalid = -1,
   Electricity = static_cast<int>FuelType::Electricity,
   NaturalGas = static_cast<int>FuelType::NaturalGas,
   DistrictHeating,
   ...
   Num
};
```

The reverse will probably work also.